src/Observers/EventDrivenObserver.cpp

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2018 United States Government as represented by the
+// Administrator of the National Aeronautics and Space Administration.
+// All Rights Reserved.
+#include "Observers/EventDrivenObserver.h"
+
+namespace PCOE {
+    EventDrivenObserver::EventDrivenObserver(MessageBus& messageBus,
+                                             std::unique_ptr<Observer>&& observer,
+                                             std::string source)
+        : bus(messageBus),
+          obs(std::move(observer)),
+          source(std::move(source)),
+          inputWatcher(bus,
+                       source,
+                       obs->getModel()->getInputs(),
+                       obs->getModel()->getInputVector()),
+          outputWatcher(bus,
+                        source,
+                        obs->getModel()->getOutputs(),
+                        obs->getModel()->getOutputVector()) {
+        Expect(obs, "Observer pointer is empty");
+        auto inputs = obs->getModel()->getInputs();
+        for (auto i : inputs) {
+            bus.subscribe(this, this->source, i);
+        }
+        auto outputs = obs->getModel()->getOutputs();
+        for (auto o : outputs) {
+            bus.subscribe(this, this->source, o);
+        }
+    }
+
+    EventDrivenObserver::~EventDrivenObserver() {
+        bus.unsubscribe(this);
+    }
+
+    void EventDrivenObserver::processMessage(const std::shared_ptr<Message>& message) {
+        latestTimestamp = seconds(message->getTimestamp());
+
+        if (inputWatcher.allPresent() && outputWatcher.allPresent()) {
```

Comment:
Is there a concern that the *watcher might not have processed this message by the time the event driven observer gets to this point? This could be a problem if we are receiving a specific message very infrequently

Reply:
Good catch. I've been going back and forth on that. As implemented, subscribers are deterministically called in the order that they subscribed, so the watcher `processMessage` is guaranteed to be called before the observer `processMessage`. As implemented right now, the whole thing is synchronous, so this also guarantees that the watcher is done.

Once I make everything async, `allPresent` and `processMessage` will share a mutex, which still leaves a race condition where execution of the watcher could be delayed such that the observer checks `allPresent` before the watcher has started processing. If we decide that that's not acceptable, I can remove the watchers from the bus and feed them messages synchronously from the observer `processMessage`.

Also note that this condition is checked on every message, so it would only be a problem if *all* messages are received very infrequently *and* the last message needed to switch `allPresent` to true is the one that hits the race condition.